game-backend/src/IntegrationTest.test.ts

```diff
@@ -0,0 +1,228 @@
+import StateBuilder from './test-helper/StateBuilder';
+import { HEALTH_LOSS_RATE } from './game-engine/mechanics/hunger-handler';
+import deepClone from './helper/deepClone';
+import MainController, { GAME_RESTART_TIME } from './MainController';
+import { TICK_INTERVAL } from './GameController';
+
+const ticksTillDeath = Math.floor(INITIAL_HEALTH / HEALTH_LOSS_RATE) + 1;
+
+const performTickNTimes = (
+  startState: State,
+  userContextContainers: UserContextContainer,
+  times: number
+): State => {
+  let state = startState;
+  for (let i = 0; i < times; i++) {
+    [state] = performTick(state, userContextContainers);
+  }
+  return state;
+};
+
+describe('Complete game-engine', () => {
+  const UNIT_ID1 = 'unit1';
+  const UNIT_ID2 = 'unit2';
+
+  const script = `
+    if (nearestFood.position.x > userUnit.position.x) {
+      move('RIGHT');
+      if (nearestFood.position.x - userUnit.position.x > 2) {
+        move('RIGHT');
+      }
+    } else if (nearestFood.position.x < userUnit.position.x) {
+      move('LEFT');
+      if (nearestFood.position.x - userUnit.position.x < -2) {
+        move('LEFT');
+      }
+    } else if (nearestFood.position.y > userUnit.position.y) {
+      move('DOWN');
+      if (nearestFood.position.y - userUnit.position.y > 2) {
+        move('DOWN');
+      }
+    } else if (nearestFood.position.y < userUnit.position.y) {
+      move('UP');
+      if (nearestFood.position.y - userUnit.position.y < -2) {
+        move('UP');
+      }
+    }
+    consume();
+  `;
+
+  let startState: State;
+  let userContextContainers: UserContextContainer;
+  beforeEach(() => {
+    startState = StateBuilder.buildState()
+      .addUnit({ id: UNIT_ID1, x: 0, y: 0 })
+      .addUnit({ id: UNIT_ID2, x: 20, y: 20 })
+      .build();
+    userContextContainers = {
+      userContextIds: [UNIT_ID1, UNIT_ID2],
+      userContextMap: {
+        [UNIT_ID1]: {
+          unitId: UNIT_ID1,
+          userScript: {
+            script: '',
+          },
+          store: {},
+        },
+        [UNIT_ID2]: {
+          unitId: UNIT_ID2,
+          userScript: {
+            script: '',
+          },
+          store: {},
+        },
+      },
+    };
+  });
+
+  it('add two units and let game end', () => {
+    let state: State = deepClone(startState);
+    state = performTickNTimes(state, userContextContainers, ticksTillDeath);
+
+    expect(getGameObjectsPerType(state, GameObjectTypes.Unit).length).toBe(0);
+  });
+
+  it('one unit eats, the other does not', () => {
+    for (const context of Object.values(userContextContainers.userContextMap)) {
+      if (context.unitId === UNIT_ID1) {
+        context.userScript = {
+          script,
+        };
+      }
+    }
+
+    let state: State = deepClone(startState);
+    state = performTickNTimes(state, userContextContainers, ticksTillDeath);
+
+    expect(getGameObjectsPerType(state, GameObjectTypes.Unit).length).toBe(1);
+  });
+
+  it('both eat but one boosted', () => {
```

Comment:
This test will probably fail regularly. (Credits to stucki)
Even failed once in the 5 times i ran them.

Reply:
It should not. We want to have a health boosted unit being able to survive longer as a non-boosted unit. In 50 matches this should hold true for at least half the matches... What do you think?

Reply:
It could hold true, but does not necessarily.
The scripts are pretty "stupid" and therefore if one unit is closer to a food, that unit always gets the food. In many cases this could be true, but it does not have to be.
Anyway whats the point of this test? what should it ensure?

Reply:
It should ensure that the boosted unit has a clear advantage over a non-boosted unit in such a scenario.

Reply:
And test things like we discovered just now regarding the immutability. (But that's secondary). 

I now added a check. where the units only consume food items and no other objects. This probably was a major factor that added randomness to the scenario. 

I realize it's not a perfect test. But in my opinion we should have the ability to test game balancing, which this test is for... 

Should we make that test go to a warning state instead?

Reply:
The advantage of one single powerup is not that big so in many cases the test will fail. (Like in my case in 1 out of 5 runs)
This test does test the basic script which takes so much randomness into account so this test is not reliable at all.
If you want to add tests for gamebalancing thats fine, but then we would need a smarter script in order to make sense.

Reply:
What if we update the test so that:
- one unit does not move at all
- one unit tries to gather food

Yes, this would not test the powerups (but as @sir-marc mentioned the current script does a poor job in testing this as we need different strategies for balancing)
More essentially, the test would be reliable because this scenario must always be valid.

Reply:
That's fair. I can update it late afternoon. I especially want to keep the part where we test it over multiple matches.